### PR TITLE
brings the proposal doc inline with actual code

### PR DESCRIPTION
**Pull request by copejon**
Signed-off-by: Jon Cope <jcope@redhat.com>

### CODE REVIEW

**Comment by copejon** [comment]
cc @jeffvance 
The diagrams still need to be reworked.

File: design/object-bucket/object-bucket.md [hunk 1]
```diff
@@ -3,57 +3,47 @@
...
+Currently, the Rook-Ceph operator enables the dynamic creation of Ceph cluster, object stores, and users via Rook native Custom Resource Definitions.  This proposal introduces the next logical step in this design: the addition a controller for handling dynamic bucket creation by leveraging the [lib-bucket-provisioner library](https://github.com/yard-turkey/lib-bucket-provisioner).  This library defines two new CRDs: ObjectBucketClaim and ObjectBucket.  Additionally, the library wraps a Kubernetes to watch related API events.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:10]: nit: "Ceph clusters..." plural

Comment (jeffvance) [reply design/object-bucket/object-bucket.md:10]: _Additionally, the library wraps a Kubernetes to watch related API events._
Additionally, the library provides, to all bucket provisioners, a Kubernetes controller to watch related API events.

File: design/object-bucket/object-bucket.md [hunk 2]
```diff
@@ -3,57 +3,47 @@
...
-- ObjectBucket: a non-namespace, admin facing Custom Resource Definition representing the fulfillment of a user request for a dynamically created bucket in an existing object store (similar in use to PersistentVolumes).  ObjectBuckets serve to provide discrete information about the bucket and object store provider to admins.
+- Admin: A Kubernetes cluster administrator with RBAC permissions to instantiate and control access to cluster storage. The admin  provisions object stores via CephObjectStore custom resources and creates StorageClasses to enable user access.  
+- User: A Kubernetes cluster user with limited permissions, typically confined to CRUD operations on Kubernetes objects within a single namespace.  The user will request Ceph buckets by instantiating ObjectBucketClaims.
+- ObjectBucketClaim: a namespaced, user facing Custom Resource Definition representing a request for a bucket.  OBCs reference a StorageClass to differentiate between provisioners, in the same manner as a PVC would.  Depending on the StorageClass, they bucket may be created dynamically or preexisting.  In the later case, new credentials are created and attached to the existing bucket.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:16]: _- ObjectBucketClaim: a namespaced, user facing Custom Resource Definition representing a request for a bucket.  OBCs reference a StorageClass to differentiate between provisioners, in the same manner as a PVC would.  Depending on the StorageClass, **they** bucket may be created dynamically or preexisting._
- ObjectBucketClaim (OBC): a namespaced, user facing Custom Resource Definition representing a request for a bucket.  OBCs reference a StorageClass to differentiate between provisioners, in the same manner as a Persistent Volume Claim (PVC) would.  Depending on the StorageClass, the bucket referenced in the OBC may get created dynamically or already exist.

File: design/object-bucket/object-bucket.md [hunk 3]
```diff
@@ -3,57 +3,47 @@
...
+- ObjectBucket: a non-namespace, admin facing CRD representing the fulfillment of a user request for a bucket, in the same manner as PV.  ObjectBuckets serve to provide discrete information about the bucket and object store provider and are not intended for Users.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:17]: _- ObjectBucket: a non-namespace, admin facing CRD representing the fulfillment of a user request for a bucket, in the same manner as PV. ObjectBuckets serve to provide discrete information about the bucket and object store provider and are not intended for Users._
- ObjectBucket (OB): a non-namespace, admin facing CRD representing the fulfillment of a user request for a bucket, in the same manner as a Persistent Volume (PV). OBs serve to provide discrete information about the bucket and object store provider and are not intended for Users.

File: design/object-bucket/object-bucket.md [hunk 4]
```diff
@@ -3,57 +3,47 @@
...
 1. Users must have permissions to PUT/GET/DELETE objects in the requested bucket.
-1. Users must have permissions to define the access policies of objects in the requested bucket.
+1. Users must have permissions to define the  policies and ACLs of objects in the requested bucket.
 1. Users must be provided credential and connection information in a Pod-consumable resource.
-1. Users must be able to reacquire bucket credential and connection information should the ObjectBucketClaim or child resource be deleted.
-1. Deleting ObjectBucketClaims and/or ObjectBuckets should not cause the deletion of a bucket to prevent accidental loss of data.
-1. The bucket naming convention, provided by the bucket lib, creates a unique bucket name (in this case by concatenating the OBC's namespace + name to form the bucket name).
-1. The object store endpoint must be exposed inside the cluster through a Service.
+1. Credential and Connection information should be protected from accidental deletion.
+1. The bucket naming convention, provided by the bucket lib, creates a unique bucket name  by concatenating the OBC's namespace + name.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:38]: are you referring the the name of the actual bucket or the OB's name here?
and I think our only requirements re. bucket naming are to provide a way to create unique bucket names and refer to existing names.

File: design/object-bucket/object-bucket.md [hunk 5]
```diff
@@ -63,7 +53,7 @@ _As an admin, I want to expose an existing Ceph object store to users so they ca
 
 ![admin-actions](./cobc-admin.png)
 
-1. The admin creates a [StorageClass](#storageclass) with the `provisioner=ceph.rook.io/object`, `parameters[serviceName]=OBJECT-STORE-SERVICE`, and `parameters[serviceNamespace]=OBJECT-STORE-NAMESPACE` key-values.
+1. The admin creates a [StorageClass](#storageclass) with the `provisioner=ceph.rook.io/object`, `parameters[objectStore]=OBJECT-STORE`, and `parameters[objectStoreNamespace]=OBJECT-STORE-NAMESPACE` key-values.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md]: Is OBJECT-STORE the name of the object store CR or it's service?

File: design/object-bucket/object-bucket.md [hunk 6]
```diff
@@ -72,40 +62,36 @@ _As an admin, I want to quickly list all dynamically created buckets in a cluste
 - The admin can get an at-a-glance picture of provisioned buckets by listing all ObjectBuckets in the cluster.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md]: Are you sure you want _dynamically created_ in the use case above? Not as easy to distinguish between OBs for dynamically created bkts vs. OBs for brownfield.

File: design/object-bucket/object-bucket.md [hunk 7]
```diff
@@ -72,40 +62,36 @@ _As an admin, I want to quickly list all dynamically created buckets in a cluste
...
 1. The operator detects the new OBC instance.
-    1. The operator checks the `provisioner` field of the OBC's StorageClass.  If it matches, continue to the next step. Else, end the process here.
-    1. The operator gets the object store service using `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
-1. The operator creates a CephObjectUser if it does not exist in order to generate the operator's access key.
-1. The operator creates the bucket using the service endpoint and access key.  The bucket name will be a concatenation of the  OBC's `metadata.namespace` and `metadata.name`.
+    1. The operator checks the `provisioner` field of the OBC's StorageClass.  If it matches, continue to the next step. Else, the process ends.
+    1. The object store service name is derived from the object
+1. The operator creates a new object store user.
+1. The operator creates the bucket using the object store endpoint and access key.
     1. If the bucket already exists in the store, the operator will instead regenerate the ConfigMap and Secret associated with that bucket.  This allows users the ability to reacquire buckets who's claims have been deleted.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md]: Is this correct?

File: design/object-bucket/object-bucket.md [hunk 8]
```diff
@@ -72,40 +62,36 @@ _As an admin, I want to quickly list all dynamically created buckets in a cluste
...
-1. The respective ConfigMap and user Secret are deleted (done by the bucket lib).
-1. The OBC is garbage collected.
-1. The OB phase is changed from "bound" to "lost"
-
-    Note: this does not cause the deletion of the OB.  Since the actual bucket is not deleted, it is reasonable that it should still tracked in Kubernetes.
+1. The user deletes the OBC
+1. The OBC is marked for deletion and left in the foreground. Through owner references, the Secret and ConfigMap are marked for deletion as well.  Finalizers are used to stall garbage collection until bucket deletion succeeds.
+1. The operator executes the Deletion method.  Depending on the StorageClass' `retainPolicy`, the bucket is either deleted or suspended.
+1. The finalizers are removed from the Secret, ConfigMap, and ObjectBucket.  The Secret and ConfigMap are garbage collected and the OBC is deleted.
+1. The operator deletes the OB.
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:118]: Do we want a brownfield use case? That would also be a place to highlight current brownfield limitations.

File: design/object-bucket/object-bucket.md [hunk 9]
```diff
@@ -72,40 +62,36 @@ _As an admin, I want to quickly list all dynamically created buckets in a cluste
...
 _As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
 
-1. The user deletes the OBC via `kubectl delete objectbucketclaim -n default FOO`.
-1. The OBC is marked for deletion and left in the foreground and the `Delete()` method, implemented by this provisioner, is invoked by the bucket lib.
-1. The respective ConfigMap and user Secret are deleted (done by the bucket lib).
-1. The OBC is garbage collected.
-1. The OB phase is changed from "bound" to "lost"
-
-    Note: this does not cause the deletion of the OB.  Since the actual bucket is not deleted, it is reasonable that it should still tracked in Kubernetes.
+1. The user deletes the OBC
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md:114]: This describes k8s cleanup operations performed by the lib but doesn't mention the steps done by the ceph provisioner.

**Comment by jeffvance** [comment]
Should there be a **Restrictions/Limitations** section?

File: design/object-bucket/object-bucket.md [hunk 10]
```diff
@@ -126,159 +112,170 @@ A [PR](https://github.com/kubernetes/kubernetes/pull/72384) exists for enabling
...
+  bucketName: BUCKET-NAME [2]
+  generateBucketName: PREFIX [3]
+  additionalConfig: [4]
+  objectBucketName: `OB-NAME` [5]
+  status:
+    phase: {"pending", "bound", "released", "failed"}  [6]  
 ```
- 
+1. `storageClassName` is used to target the desired Object Store.  Used by the operator to get the Object Store service URL.
+1. `bucketName` is the desired name of the ceph bucket.  Mutually exclusive with `generateBucketName`
```
Comment (jeffvance) [inline design/object-bucket/object-bucket.md]: Important to note that both names are omitted for brownfield...

File: design/object-bucket/object-bucket.md (hunk 5 above)
Comment (copejon) [reply design/object-bucket/object-bucket.md]: the object store.  this was a change we made a while ago.  the service name is the same as the store.  From UX PoV, it makes more sense to me to have admins provide the object store.

File: design/object-bucket/object-bucket.md (hunk 7 above)
Comment (copejon) [reply design/object-bucket/object-bucket.md]: nope, will remove

File: design/object-bucket/object-bucket.md (hunk 1 above)
Comment (jeffvance) [reply design/object-bucket/object-bucket.md:10]: comment still applies. "cluster" should be plural or write "a cluster..."

**Pull request closed without merging**